---
title: "GWAS tutorial"
author: "Eryn McFarlane"
date: "`r Sys.Date()`"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

### Learning outcomes of GWAS tutorial
1) Understand what a GWAS is, and how to run it in GEMMA
2) Think about specific predictions for GWAS, and what we really expect to see for most of them
3) Think critically about when GWAS are robust, and when they should be taken skeptically

```{r simulations}
### This is from Alex Buerkle's Population Genetics lab notes
###simulate some data
set.seed(42)
nloci<-10000  ## we will have data from nloci
nind<-2000  ## we will have data from nind that were sampled from the true population
sim.theta<-5  # high parameter means high polymorphism, must be positive and > zero, can be thought of as analygous to nucleotide polymorphism

## simulate allele frequencies at nloci number of loci, by random draws from a beta
sim.p<-rbeta(nloci, sim.theta, sim.theta)
hist(sim.p, breaks=seq(0,1,0.1))

# simulate genotypes in the sample
sim.x <- matrix(rbinom(nloci*nind, 2, prob=sim.p), nrow=nind, ncol=nloci) ### this gives us individuals in rows and snps in columns
str(sim.x)
### this is like the .ped set up, but not entirely. @ERYN - go back to this to make it the bimbam right away so it fits in GEMMA from the start.

###some things to note - these are assumed to be independent. This isn't realistic, there would certainly be LD
### I haven't simulated chromosomes, just allele frequencies

### one more thing to simulate is the phenotypes, and the number of markers that they're based one.

### let's do one that's categorical and based on one SNP, like the horn traits in soay sheep

y_cat<-as.factor(ifelse((5+0.3*sim.x[,15]+rnorm(2000, 0, 1))<4, 'no horn', ifelse((5+0.3*sim.x[,15]+rnorm(2000, 0, 1))>6.5, 'big horn', 'small horn'))) ### this is just a linear relationship between SNP 15 and this variable, plus error, that I've then put into categories

y_quant<-0.1*sim.x[,10]+0.1*sim.x[,20]+0.1*sim.x[,30]+0.1*sim.x[,40]-0.1*sim.x[,50]-0.1*sim.x[,60]-0.1*sim.x[,70]+0.1*sim.x[,80]+0.1*sim.x[,90]+0.1*sim.x[,100]+rnorm(2000, 0, 5)

hist(y_quant)

y_polygenic<-0.0001*sim.x[,11]+
   0.0001*sim.x[,21]+ 
  0.0001*sim.x[,31]+ 
  0.0001*sim.x[,41]+ 
  0.0001*sim.x[,51]+ 
  0.0001*sim.x[,61]+ 
  0.0001*sim.x[,71]+ 
  0.0001*sim.x[,81]+ 
  0.0001*sim.x[,91]+ 
  0.0001*sim.x[,101]+ 
  0.0001*sim.x[,111]+ 
  0.0001*sim.x[,121]+ 
  0.0001*sim.x[,131]+ 
  0.0001*sim.x[,141]+ 
  0.0001*sim.x[,151]+ 
  0.0001*sim.x[,161]+ 
  0.0001*sim.x[,171]+ 
  0.0001*sim.x[,181]+ 
  0.0001*sim.x[,191]+ 
  0.0001*sim.x[,201]+ 
  0.0001*sim.x[,211]+ 
  0.0001*sim.x[,221]+ 
  0.0001*sim.x[,231]+ 
  0.0001*sim.x[,241]+ 
  0.0001*sim.x[,251]+ 
  0.0001*sim.x[,261]+ 
  0.0001*sim.x[,271]+ 
  0.0001*sim.x[,281]+ 
  0.0001*sim.x[,291]+ 
  0.0001*sim.x[,301]+ 
  0.0001*sim.x[,311]+ 
  0.0001*sim.x[,321]+ 
  0.0001*sim.x[,331]+ 
  0.0001*sim.x[,341]+ 
  0.0001*sim.x[,351]+ 
  0.0001*sim.x[,361]+ 
  0.0001*sim.x[,371]+ 
  0.0001*sim.x[,381]+ 
  0.0001*sim.x[,391]+ 
  0.0001*sim.x[,401]+ 
  0.0001*sim.x[,411]+ 
  0.0001*sim.x[,421]+ 
  0.0001*sim.x[,431]+ 
  0.0001*sim.x[,441]+ 
  0.0001*sim.x[,451]+ 
  0.0001*sim.x[,461]+ 
  0.0001*sim.x[,471]+ 
  0.0001*sim.x[,481]+ 
  0.0001*sim.x[,491]+ 
  0.0001*sim.x[,501]+ 
  0.0001*sim.x[,511]+ 
  0.0001*sim.x[,521]+ 
  0.0001*sim.x[,531]+ 
  0.0001*sim.x[,541]+ 
  0.0001*sim.x[,551]+ 
  0.0001*sim.x[,561]+ 
  0.0001*sim.x[,571]+ 
  0.0001*sim.x[,581]+ 
  0.0001*sim.x[,591]+ 
  0.0001*sim.x[,601]+ 
  0.0001*sim.x[,611]+ 
  0.0001*sim.x[,621]+ 
  0.0001*sim.x[,631]+ 
  0.0001*sim.x[,641]+ 
  0.0001*sim.x[,651]+ 
  0.0001*sim.x[,661]+ 
  0.0001*sim.x[,671]+ 
  0.0001*sim.x[,681]+ 
  0.0001*sim.x[,691]+ 
  0.0001*sim.x[,701]+ 
  0.0001*sim.x[,711]+ 
  0.0001*sim.x[,721]+ 
  0.0001*sim.x[,731]+ 
  0.0001*sim.x[,741]+ 
  0.0001*sim.x[,751]+ 
  0.0001*sim.x[,761]+ 
  0.0001*sim.x[,771]+ 
  0.0001*sim.x[,781]+ 
  0.0001*sim.x[,791]+ 
  0.0001*sim.x[,801]+ 
  0.0001*sim.x[,811]+ 
  0.0001*sim.x[,821]+ 
  0.0001*sim.x[,831]+ 
  0.0001*sim.x[,841]+ 
  0.0001*sim.x[,851]+ 
  0.0001*sim.x[,861]+ 
  0.0001*sim.x[,871]+ 
  0.0001*sim.x[,881]+ 
  0.0001*sim.x[,891]+ 
  0.0001*sim.x[,901]+ 
  0.0001*sim.x[,911]+ 
  0.0001*sim.x[,921]+ 
  0.0001*sim.x[,931]+ 
  0.0001*sim.x[,941]+ 
  0.0001*sim.x[,951]+ 
  0.0001*sim.x[,961]+ 
  0.0001*sim.x[,971]+ 
  0.0001*sim.x[,981]+ 
  0.0001*sim.x[,991]+ 
  0.0001*sim.x[,1001]+rnorm(2000, 0, 5)


### need to output something that looks like a bimbam datafiles

```
```{r choose your own adventure}
set.seed() ### put in your favourite number here!

test_individuals<-sample(1:2000, 1000, replace=FALSE)
test_snps<-sample(1:10000, 5000, replace=FALSE)

write.table(y_cat[test_individuals], row.names=FALSE, col.names=FALSE, file="y_cat.txt")
write.table(y_quant[test_individuals], row.names=FALSE, col.names=FALSE, file="y_quant.txt")
write.table(y_polygenic[test_individuals], row.names=FALSE, col.names=FALSE, file="y_polygenic.txt")

SNPs<-t(sim.x[test_individuals, test_snps])
SNPs_bimbam<- cbind(1:length(rowMeans(SNPs)), rowMeans(SNPs), rowMeans(SNPs), SNPs) ### this is what makes it into a bimbam file we need
  
write.table(SNPs_bimbam, row.names = FALSE, col.names = FALSE, file="SNPs.txt")

```

```{r basic gemma, echo=FALSE}

library(coda)
library(ggplot2)

### quantitative trait
  system("~/gemma.macosx -g SNPs.txt -p y_quant.txt -gk 1 -o relmatrix", wait=TRUE)
  system("~/gemma.macosx -g SNPs.txt -p y_quant.txt -k output/relmatrix.cXX.txt -bslmm 1 -s 5000 -w 5000 -o y_quant", wait=TRUE)
  


  hyp.params <- read.table("output/y_quant.hyp.txt", header=TRUE)
  converged <- ifelse(heidel.diag(hyp.params$pve)[1,4]==1, "converged", "not_converged")
  pve<-c("PVE", mean(hyp.params$pve),quantile(hyp.params$pve, probs=c(0.5,0.025,0.975)))
  ## pge -> proportion of genetic variance explained by major effect loci
  pge <- c("PGE",mean(hyp.params$pge),quantile(hyp.params$pge, probs=c(0.5,0.025,0.975)))
  ## n.gamma -> number of variants with major effect
  n.gamma <- c("n.gamma",mean(hyp.params$n_gamma),quantile(hyp.params$n_gamma, probs=c(0.5,0.025,0.975)))

  params <- read.table("output/y_quant.param.txt",header=T,sep="\t")
  params["eff"]<-abs(params$beta*params$gamma) # add sparse effect size (= beta * gamma) to data frame - need to understand why this is as it is!

params.effects<-params[params$eff>0,]

# show number of variants with measurable effect
nrow(params.effects)


params.effects.sort<-params.effects[order(-params.effects$eff),]


# show top 10 variants with highest effect
head(params.effects.sort, 11) 

top01<-params.effects.sort[params.effects.sort$eff>quantile(params.effects.sort$eff,0.999),]

# sort variants by descending PIP
params.pipsort<-params[order(-params$gamma),]

# Show top 10 variants with highest PIP
head(params.pipsort,13)
plot<-ggplot(params, aes(as.factor(rs), gamma))+geom_jitter(size=3)+theme_bw()+geom_hline(yintercept=0.10, linetype='dashed', colour="grey")+theme(legend.position='none')+xlab("SNPs")+ylab("PIP")+theme(axis.title=element_text(size=14))

### do something so that this goes all the way up to PIP 1 - as of right now, none of these are over the threshold!

plot+geom_text(aes(label=ifelse(gamma>0.15, rs, "")), hjust="inward",vjust="inward", size=5, nudge_x = 0.5)
 


png(file="PIP_manhattenplot_190421.png", width=6, height=3, units="in", res=300)
plot_noFEs<-ggplot(params_noFEs, aes(as.factor(chr), gamma))+geom_jitter(aes(colour=as.factor(chr)), size=3)+scale_colour_manual(values=rep(c("grey", "skyblue"), 22))+theme_bw()+geom_hline(yintercept=0.10, linetype='dashed', colour="grey")+theme(legend.position='none')+xlab("Cervus linkage groups")+ylab("PIP")+theme(axis.title=element_text(size=14))
plot_noFEs
dev.off()



```

